node/src/components/network/tests2.rs

```diff
@@ -0,0 +1,240 @@
+
+    // Port for first node, other will connect to it.
+    let first_node_port = testing::unused_port_on_localhost() + 1;
+
+    let mut net = testing::network::Network::<LoadTestingReactor>::new();
+    let chainspec = Chainspec::random(&mut rng);
+
+    // Create the root node.
+    let cfg = TestReactorConfig {
+        chainspec: chainspec.clone(),
+        network_config: crate::components::network::Config::default_local_net_first_node(
+            first_node_port,
+        ),
+    };
+
+    net.add_node_with_config(cfg, &mut rng).await.unwrap();
+
+    // Create `node_count-1` additional node instances.
+    for _ in 1..node_count {
+        let cfg = TestReactorConfig {
+            chainspec: chainspec.clone(),
+            network_config: crate::components::network::Config::default_local_net(first_node_port),
+        };
+
+        net.add_node_with_config(cfg, &mut rng).await.unwrap();
+    }
+
+    info!("Network setup, waiting for discovery to complete");
+    net.settle_on(&mut rng, network_online, timeout).await;
+    info!("Discovery complete");
+
+    // At this point each node has at least one other peer. Assuming no split, we can now start
+    // gossiping a large payloads. We gossip one on each node.
+    let node_ids: Vec<_> = net.nodes().keys().cloned().collect();
+    for (index, sender) in node_ids.iter().enumerate() {
+        let dummy_payload = DummyPayload::random_with_size(&mut rng, large_size);
+
+        // Calling `broadcast_message` actually triggers libp2p gossping.
+        net.process_injected_effect_on(sender, |effect_builder| {
+            effect_builder
+                .broadcast_message(dummy_payload.clone())
+                .ignore()
+        })
+        .await;
+
+        info!(?sender, payload = %dummy_payload, round=index, total=node_ids.len(),
+              "Started broadcast/gossip of payload, waiting for all nodes to receive it");
+        net.settle_on(
+            &mut rng,
+            others_received(dummy_payload.id(), sender.clone()),
+            timeout,
+        )
+        .await;
+        info!(?sender, "Completed gossip test for sender")
+    }
+}
+
+/// Checks if all nodes are connected to at least one other node.
+pub fn network_online(
+    nodes: &HashMap<NodeId, Runner<ConditionCheckReactor<LoadTestingReactor>>>,
+) -> bool {
+    assert!(
+        nodes.len() >= 2,
+        "cannot check for an online network with less than 3 nodes"
+    );
+
+    let k_value = usize::from(K_VALUE);
+
+    // Sanity check of K_VALUE.
+    assert!(
+        k_value >= 7,
+        "K_VALUE is really small, expected it to be at least 7"
+    );
+
+    // The target of known nodes to go for. This has a hard bound of `K_VALUE`, since if all nodes
+    // end up in the same bucket, we will start evicting them. In general, we go for K_VALUE/2 for
+    // reasonable interconnection, or the network size - 1, which is another bound.
+    let known_nodes_target = (k_value / 2).min(nodes.len() - 1);
```

Comment:
For larger networks, this number seems pretty small.  As you mentioned - in a very badly imbalanced network, all nodes could end up in a single k-bucket's address space (although this should be _highly_ improbable).  But in that case, every peer should eventually be connected to `k` peers I believe.

So, I'd suggest at least changing `(k_value / 2)` to just `k_value`.

Reply:
I'm following up on this in a separate PR.